amethyst_tiles/src/map.rs

```diff
@@ -0,0 +1,365 @@
+#![allow(unused_variables)]
+
+use crate::CoordinateEncoder;
+use amethyst_assets::Handle;
+use amethyst_core::{
+    ecs::{Component, HashMapStorage, World},
+    math::{Matrix4, Point3, Vector3},
+};
+use amethyst_rendy::{palette::Srgba, SpriteSheet};
+
+/// Trait providing generic rendering functionality to all tiles. Using a tilemap requires you to provide a `Tile` type,
+/// which must implement this trait to provide the `RenderPass` with the appropriate sprite and tint values.
+pub trait Tile: 'static + Send + Sync + Default {
+    /// Takes an immutable reference to world to process this sprite and return its sprite.
+    fn sprite(&self, coordinates: Point3<u32>, world: &World) -> Option<usize> {
+        None
+    }
+
+    /// Takes an immutable reference to world to process this sprite and return its tint.
+    fn tint(&self, coordinates: Point3<u32>, world: &World) -> Srgba {
+        Srgba::new(1.0, 1.0, 1.0, 1.0)
+    }
+}
+
+/// Trait for providing access to an underlying storage type of a 3-dimensional Tile data. This is abstracted to provide
+/// for allowing more underlying storage types in the future beyond a flat array, such as networking, chunking, etc.
+pub trait Map {
+    /// The world-space (Amethyst) dimensions of a single tile in this map space (1x1x1). This is used to scale our
+    /// sprites to the world coordinate space. This should usually be the tile sprite dimensions. Beware, Z-size is taken
+    /// into consideration as well so you will usually want to se Z-size to 1.
+    fn tile_dimensions(&self) -> &Vector3<u32>;
```

Comment:
those should probably return copies. The nalgebra basic types are `Copy`. I don't have a strong opinion on this though, just worth thinking about maybe.

Reply:
Yes, performance wise its similar. I determined to stick with references due to their immutable nature - i wanted the return value to be default-immutable and force a user copy if they want to modify it.

Reply:
Yeah, let's leave it as is now.